lib/ddtrace/contrib/rest_client/request_patch.rb

```diff
@@ -0,0 +1,84 @@
+require 'ddtrace/ext/net'
+require 'ddtrace/ext/distributed'
+
+module Datadog
+  module Contrib
+    module RestClient
+      # RestClient RequestPatch
+      module RequestPatch
+        REQUEST_TRACE_NAME = 'rest_client.request'.freeze
+
+        def self.included(base)
+          base.prepend(InstanceMethods)
+        end
+
+        # InstanceMethods - implementing instrumentation
+        module InstanceMethods
+          def execute(&block)
+            datadog_trace_request do |span|
+              datadog_propagate!(span.context) if datadog_configuration[:distributed_tracing] && datadog_pin.tracer.enabled
+
+              super(&block)
+            end
+          end
+
+          protected
+
+          def datadog_propagate!(context)
+            processed_headers[Ext::DistributedTracing::HTTP_HEADER_TRACE_ID] = context.trace_id.to_s
+            processed_headers[Ext::DistributedTracing::HTTP_HEADER_PARENT_ID] = context.span_id.to_s
+            if context.sampling_priority
+              processed_headers[Ext::DistributedTracing::HTTP_HEADER_SAMPLING_PRIORITY] = context.sampling_priority.to_s
+            end
+          end
+
+          def datadog_tag_request(span)
+            span.resource = method.to_s.upcase
+            span.span_type = Ext::HTTP::TYPE
+            span.set_tag(Ext::HTTP::URL, uri.path)
+            span.set_tag(Ext::HTTP::METHOD, method.to_s.upcase)
+            span.set_tag(Ext::NET::TARGET_HOST, uri.host)
+            span.set_tag(Ext::NET::TARGET_PORT, uri.port)
+          end
+
+          def datadog_trace_request
+            span = datadog_pin.tracer.trace(REQUEST_TRACE_NAME, service: datadog_pin.service_name)
```

Comment:
Why do we use this asynchronous form instead of block form? The block form should cover most of what you're doing here implicitly, and permit you to do the customizations you require.

Reply:
Indeed block form would have been nicer, however the problem is with 
```ruby
rescue ::RestClient::ExceptionWithResponse => e
...
raise e
```
Since in block form this will be traced as error regardless of the error code.

Reply:
Could you rescue that error inside the block form, the set the error or re-raise if necessary? We do [something like this in Rails](https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/contrib/rails/middlewares.rb). It's not exactly the same, but maybe there's some inspiration to draw from it.

Reply:
I need to re-raise this error not to break any one use of rest-client.
We could modify `tracer.trace do ... end` to allow it to somehow skip `span.set_error(e)` step.

But so far I don't have a good idea how to approach this.